include/grpc++/channel_filter.h

```diff
@@ -0,0 +1,211 @@
+ *     * Neither the name of Google Inc. nor the names of its
+ * contributors may be used to endorse or promote products derived from
+ * this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ */
+
+#ifndef GRPCXX_CHANNEL_FILTER_H
+#define GRPCXX_CHANNEL_FILTER_H
+
+#include <grpc/grpc.h>
+
+#include <vector>
+
+#include "src/core/lib/channel/channel_stack.h"
+#include "src/core/lib/surface/channel_init.h"
+
+//
+// An interface to define filters.
+//
+// To define a filter, implement a subclass of each of CallData and
+// ChannelData.  Then register the filter like this:
+//   RegisterChannelFilter<MyChannelDataSubclass, MyCallDataSubclass>(
+//       "name-of-filter", GRPC_SERVER_CHANNEL, INT_MAX);
+//
+
+namespace grpc {
+
+// Represents call data.
+// Note: Must be copyable.
+class CallData {
+ public:
+  // Do not override the destructor.  Instead, put clean-up code in the
+  // Destroy() method.
+  virtual ~CallData() {}
+
+  virtual void Destroy() {}
+
+  virtual void StartTransportStreamOp(
+      grpc_exec_ctx *exec_ctx, grpc_call_element *elem,
+      grpc_transport_stream_op *op);
+
+  virtual void SetPollsetOrPollsetSet(
+      grpc_exec_ctx *exec_ctx, grpc_call_element *elem,
+      grpc_polling_entity *pollent);
+
+  virtual char* GetPeer(grpc_exec_ctx *exec_ctx, grpc_call_element *elem);
+
+ protected:
+  CallData() {}
+};
+
+// Represents channel data.
+// Note: Must be copyable.
+class ChannelData {
+ public:
+  // Do not override the destructor.  Instead, put clean-up code in the
+  // Destroy() method.
+  virtual ~ChannelData() {}
+
+  virtual void Destroy() {}
+
+  virtual void StartTransportOp(
+      grpc_exec_ctx *exec_ctx, grpc_channel_element *elem,
+      grpc_transport_op *op);
+
+ protected:
+  ChannelData() {}
+};
+
+namespace internal {
+
+// Defines static members for passing to C core.
+template<typename ChannelDataType, typename CallDataType>
+class ChannelFilter {
+  static const size_t call_data_size = sizeof(CallDataType);
+
+  static void InitCallElement(
+      grpc_exec_ctx *exec_ctx, grpc_call_element *elem,
+      grpc_call_element_args *args) {
+    CallDataType* call_data = elem->call_data;
```

Comment:
This method should be just:

```c
new (elem->call_data) CallDataType();
```

... there's no guarantee that the data pointed to is initialized, so we want to avoid calling the destructor (or that might crash)